Replace growl errors by modal in VBA flow

aldo-expensify opened the pull request:
Replacing all growls in the VBA flow by the modal.

The error message coming from the backend is being shown in the ConfirmModal now.

Errors messages associated with inputs where we don't handle highlighting yet, like checkboxes, are shown in the ConfirmModal.

Error modal generic message: 'Please double check any highlighted field and try again'

### Fixed Issues

$ https://github.com/Expensify/App/issues/4785

### Tests / QA Steps

Company Step: /bank-account/company

1. Navigate to `/bank-account/company`
2. Fill all inputs, input an invalid 'Zip code' (less than 6 numbers)
3. Click 'Save & continue'
4. Expected results: errors in inputs validated in the front (i.e. zip code, tax id, etc) should be highlighted in the form and a generic message should be shown in a modal.<img width="889" alt="Screen Shot 2021-09-01 at 3 56 23 PM" src="https://user-images.githubusercontent.com/87341702/131756105-2d9dab69-9201-4ff0-a042-ffaaa0896905.png">
5. Input a 'Zip Code' of 6 numbers (i.e. `123123`) and input an invalid phone number (i.e. `123`).
6. Click 'Save & continue'
7. Expected results: The phone number is a field validated in the backend, we will display a specific message in the error modal.<img width="889" alt="Screen Shot 2021-09-01 at 3 57 39 PM" src="https://user-images.githubusercontent.com/87341702/131756208-19788fd7-dbd3-4bda-8d1b-32708a9d2ced.png">


ACH Contract / Beneficial Owners Step: /bank-account/contract

1. Navigate to `/bank-account/contract`
2. Check `I accept the terms and conditions` and `I certify that the information provided is true and accurate`
3. Check `Somebody else owns more than 25% of XXX`, this will add a form to input a beneficial owner's information. Don't fill any of the beneficial owner data.
4. Click 'Save & continue'
5. Expected result: Highlighted invalid field (frontend validation) with error modal with generic message <img width="889" alt="Screen Shot 2021-09-01 at 4 20 07 PM" src="https://user-images.githubusercontent.com/87341702/131757790-70007c8c-7725-44ae-aea4-c9e4d0fecca0.png">

Requestor Step: /bank-account/requestor

1. Navigate to `/bank-account/requestor`
2. Fill all inputs
3. Input a date with invalid format in 'Date of birth', in example: `asd`
4. Expected result: 'Date of birth' field is highlighted with an error, we get the error modal with the generic message (this validation happened in the frontend).<img width="889" alt="Screen Shot 2021-09-01 at 4 09 26 PM" src="https://user-images.githubusercontent.com/87341702/131757046-0bdb010e-8d11-4b7c-abd5-a93fdd0d8108.png">
5. Input a date rejected by the backend in the 'Date of birth' , in example: `2020-10-10`
6. Expected result: The error modal displays a specific error about the 'Date of birth' being incorrect. No field is highlighted.<img width="889" alt="Screen Shot 2021-09-01 at 4 12 51 PM" src="https://user-images.githubusercontent.com/87341702/131757305-a35a71f8-8dd8-4909-8085-9dd716c4c21d.png">



### Tested On

- [x] Web
- [x] Mobile Web
- [ ] Desktop
- [ ] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->

#### Android
<!-- Insert screenshots of your changes on the Android platform-->

aldo-expensify committed:
Move error modal to ReimbursementAccountPage

aldo-expensify committed:
Pass strictly true false to ConfirmModal.isVisible

aldo-expensify committed:
Replace growls with errors by modal

aldo-expensify committed:
Replace growls. Refactor identity validation.

Split setErrorModalVisible into two methods.

aldo-expensify committed:
Revert ValidationUtils identity validation

marcaaron commented on src/pages/ReimbursementAccount/RequestorStep.js:
```diff
@@ -71,17 +71,18 @@ class RequestorStep extends React.Component {
             ssnLast4: this.state.ssnLast4,
-        })) {
+        };
+        if (!isValidIdentity(identity)) {
```
We can undo this change for now. Let's only make the changes we need to?

aldo-expensify replied:
done

marcaaron commented on src/pages/ReimbursementAccount/ReimbursementAccountPage.js:
```diff
@@ -238,6 +239,14 @@ class ReimbursementAccountPage extends React.Component {
                     )}
+                    <ConfirmModal
+                        title={this.props.translate('companyStep.confirmModalTitle')} // TODO: Update translation key
+                        onConfirm={() => hideErrorModal()}
```
No need to use an arrow function for this can just do `onConfirm={hideErrorModal}`

aldo-expensify replied:
done

marcaaron commented on src/pages/ReimbursementAccount/ReimbursementAccountPage.js:
```diff
@@ -238,6 +239,14 @@ class ReimbursementAccountPage extends React.Component {
+                        title={this.props.translate('companyStep.confirmModalTitle')} // TODO: Update translation key
+                        onConfirm={() => hideErrorModal()}
+                        prompt={this.props.reimbursementAccount.errorModalMessage || this.props.translate('companyStep.confirmModalPrompt')}
+                        isVisible={Boolean(this.props.reimbursementAccount.isErrorModalVisible)}
```
Any reason to do a `Boolean()` cast here?

aldo-expensify replied:
The property `this.props.reimbursementAccount.isErrorModalVisible` is undefined at the start, so you get some errors in the console if you pass undefined to `isVisible'

![Screen Shot 2021-08-30 at 4 07 43 PM](https://user-images.githubusercontent.com/87341702/131416966-4f7763c7-3492-41a3-a334-94e96c7048fe.png)

marcaaron replied:
Ah maybe give it a default value? If it's `undefined` then pass `false` ?

aldo-expensify replied:
you mean like `isVisible={this.props.reimbursementAccount.isErrorModalVisible || false}`, or setting the default as default prop in ConfirmModal.js? If it is the second, the linter complains because it doesn't like default props for 'required props'. If it is the first... I guess it is almost the same.

Linter: `propType "isVisible" is required and should not have a defaultProps declaration`

marcaaron replied:
>If it is the first... I guess it is almost the same.

Fair enough, I've seen it both ways, but mostly have seen use of `lodashGet()` when needing a default value.

aldo-expensify replied:
> > If it is the first... I guess it is almost the same.
> 
> Fair enough, I've seen it both ways, but mostly have seen use of `lodashGet()` when needing a default value.

Hmm you are right, lodashGet seems to be used a lot for this in the project... I guess I'll be consistent

marcaaron replied:
Ok I could call this a NAB (not a blocker) too - seems minor looking again today ;)

marcaaron commented on src/libs/ValidationUtils.js:
```diff
@@ -75,32 +75,38 @@ function isValidSSNLastFour(ssnLast4) {
     }
 
     if (!isValidSSNLastFour(identity.ssnLast4)) {
         showBankAccountFormValidationError(translateLocal('bankAccount.error.ssnLast4'), true);
```
We are still passing the `shouldGrowl` argument in all of these.

aldo-expensify replied:
ops, good catch

marcaaron commented on src/libs/actions/BankAccounts.js:
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
+ * @param {String} errorMessage The error message to be displayed in the modal's body. We show translate('companyStep.confirmModalPrompt') if not provided.
+ */
+
+function showErrorModal(errorMessage = null) {
```
Should maybe be more specific for now `showBankAccountErrorModal()` ?

aldo-expensify replied:
I have my doubts about that. I feel like the developer can easily know that this function was imported from BankAccount and I like to avoid having to include that in the methods we export, but if you think is better to change it because it can avoid some confusion later.. I can do it anyway

marcaaron replied:
If you were deciding to use this pattern in other places then you'd have a generic `showErrorModal()` method when in reality it only relates to a single API command. So, I think it's better to be specific, but if you'd prefer to wait for that to become a problem them I'd understand. If you strongly disagree then disagree and we can keep discussing it.

aldo-expensify replied:
I don't strongly disagree, I'll rename it

marcaaron commented on src/libs/actions/BankAccounts.js:
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
+ */
+
+function showErrorModal(errorMessage = null) {
+    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage: errorMessage});
```
NAB, if you make the argument match the property name we can do this

```js
Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage});
```

aldo-expensify replied:
Done

marcaaron commented on src/libs/actions/BankAccounts.js:
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
+/**
+ * Hide error modal
+ */
+function hideErrorModal() {
```
`hideBankAccountErrorModal()`?

marcaaron reviewed (commented):
Good start. I left a few ideas for improvements.

aldo-expensify committed:
Remove unnecesary arrow function

aldo-expensify committed:
Revert unnecesary change

aldo-expensify committed:
Removed unexisting param on function call

aldo-expensify committed:
Parameter rename in showErrorModal

aldo-expensify committed:
Remove unnecessary comment

aldo-expensify committed:
Rename methods to hide/show VBA error modal

aldo-expensify committed:
Handle error modal visibility default using lodash

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_replace-growl-by-modal

aldo-expensify commented on src/libs/actions/BankAccounts.js:
```diff
@@ -603,16 +603,29 @@ function validateBankAccount(bankAccountID, validateCode) {
- * Set the current error message. Show Growl for errors which are not yet handled by the error Modal.
+ * Show error modal and optionally a specific error message
+ *
+ * @param {String} errorMessage The error message to be displayed in the modal's body. We show translate('companyStep.confirmModalPrompt') if not provided.
```
Removed

marcaaron reviewed (commented):
Changes look perfect! 🏆 

I think the one thing left to resolve is that there are some places where the confirm modal title does not entirely make sense. Trying to get a reply to the thread here [now](https://expensify.slack.com/archives/C020EPP9B9A/p1630350084180500?thread_ts=1630092331.070800&cid=C020EPP9B9A)

aldo-expensify commented:
> think the one thing left to resolve is that there are some places where the confirm modal title does not entirely make sense. Trying to get a reply to the thread here [now](https://expensify.slack.com/archives/C020EPP9B9A/p1630350084180500?thread_ts=1630092331.070800&cid=C020EPP9B9A)

To handle this, would it be fine to include an extra parameter on the `showBankAccountErrorModal` like:

```
function showBankAccountErrorModal(errorModalMessage = null, errorModalTitle = null) {
    Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {isErrorModalVisible: true, errorModalMessage, errorModalTitle});
}
```
Use it in the display modal like:
```
<ConfirmModal
  title={lodashGet(this.props, 'reimbursementAccount.errorModalTitle', this.props.translate('companyStep.confirmModalTitle'))}
```

Then, when we set the error from the server and want a custom title, we do:

```
showBankAccountErrorModal(error, translateLocal('bankAccount.error.serverErrorModalTitle'));
```

marcaaron commented:
>To handle this, would it be fine to include an extra parameter on the showBankAccountErrorModal
>Then, when we set the error from the server and want a custom title

Let's wait for feedback first? I haven't seen much discussion about whether these particular errors should have custom titles.

marcaaron commented:
>Hey @marcaaron, considering we are going to use the Oops header, does the proposed way of handling it above work for you?

Ah no, I think we should make all the headers say "Oops" if that's acceptable? 
I'm not sure there is a need to get very custom on these headers and it is always best to do the simplest thing.

aldo-expensify commented:
> Ah no, I think we should make all the headers say "Oops" if that's acceptable?
> I'm not sure there is a need to get very custom on these headers and it is always best to do the simplest thing.

Okk, I think I was miss-understanding  that we wanted different headers. I will change the generic header of the error modal to "Oops" then!

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_replace-growl-by-modal

aldo-expensify committed:
Correct mistake when merging conflicts

aldo-expensify committed:
Replace error modal header with Oops

aldo-expensify commented:
Update:
- Replaced error modal header with "Oops" (EN) / "Ups" (ES)
- Moved translation keys related to error modal from `companyStep` into `bankAccount`

marcaaron reviewed (changes requested):
Looks perfect there are just two small issues where unnecessary new lines are added. But other than than great work.

aldo-expensify committed:
Remove unnecessary empty lines

aldo-expensify commented on src/libs/ValidationUtils.js:
```diff
@@ -74,33 +74,39 @@ function isValidSSNLastFour(ssnLast4) {
     return true;
 }
 
+
```
Done

aldo-expensify commented on src/libs/actions/BankAccounts.js:
```diff
@@ -612,16 +612,29 @@ function validateBankAccount(bankAccountID, validateCode) {
+ *
+ * @param {String} errorModalMessage The error message to be displayed in the modal's body.
+ */
+
```
Done

aldo-expensify commented:
Update: removed empty lines

marcaaron reviewed (approved):
Looks good! Thanks for the changes!

Approving the code. But we still need to clean up the PR description because there is nothing listed for QA Steps. We need to put something there so that QA knows what to test (and also figure out what QA is able to test or not). Otherwise, they will inevitably ping you (or me since I reviewed) and come asking if there is QA to do or not :)

Also, we should test this on all platforms to make sure the changes are safe. There's checkboxes listed so that we test on every platform. But if there's a good reason not to then we can mention it somewhere.

MitchExpensify commented:
Flo is likely offline for the night. Anyone else we can request? @marcaaron

aldo-expensify commented:
> Looks good! Thanks for the changes!
> 
> Approving the code. But we still need to clean up the PR description because there is nothing listed for QA Steps. We need to put something there so that QA knows what to test (and also figure out what QA is able to test or not). Otherwise, they will inevitably ping you (or me since I reviewed) and come asking if there is QA to do or not :)
> 
> Also, we should test this on all platforms to make sure the changes are safe. There's checkboxes listed so that we test on every platform. But if there's a good reason not to then we can mention it somewhere.

I'll start cleaning this up now (QA steps, outdated pictures), thanks for the reminder!

marcaaron commented:
>There is one thing in the bank-account/contract step that looks like a bug, but it could also be that it just doesn't work well in the development environment. I'll test it in main branch, it feels unrelated to what we are doing.

Nice! That was just called out by @JmillsExpensify [here](https://expensify.slack.com/archives/C020EPP9B9A/p1630536226414100?thread_ts=1630532268.395300&cid=C020EPP9B9A). I think we'll need to create a follow up issue to look into it.

If anything else pops up that you notice we can check in the free plan channel to see if @MitchExpensify & @kevinksullivan are aware of the issue yet.

MitchExpensify commented:
Nice one @aldo-expensify! 

I'll create a follow up issue to change the error copy we use when the age entered is under our required 18 years old

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1192328137) to staging by @marcaaron in version: 1.0.91-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1196332434) to production by @roryabraham in version: 1.0.92-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
